### ISSUE TITLE
Infinite loop on upload retry functionality

### ISSUE THREAD

**Issue by directionless** [logstash-plugins/logstash-output-s3#160]
Hi!  I'm trying to get the s3 output plugin working, and I ran into what looks like an infinite loop.

My logs are emitting a huge rate of this message (slightly redacted):
```
[2017-10-13T19:09:24,271][ERROR][logstash.outputs.s3      ] Uploading failed, retrying {:exception=>Errno::ENOENT, :message=>"No such file or directory - No such file or directory - /tmp/logstash/GUID/myprefix/ls.s3.GUID.2017-10-13T18.54.part0.txt", :path=>"/tmp/logstash/GUID/myprefix/ls.s3.GUID.2017-10-13T18.54.part0.txt", :backtrace=>["org/jruby/RubyFileTest.java:239:in `size'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/file_uploader.rb:31:in `upload'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/object.rb:251:in `upload_file'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.11/lib/logstash/outputs/s3/uploader.rb:38:in `upload'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.11/lib/logstash/outputs/s3/uploader.rb:29:in `upload_async'", "org/jruby/RubyProc.java:281:in `call'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/concurrent-ruby-1.0.5-java/lib/concurrent/executor/java_executor_service.rb:94:in `run'", "Concurrent$$JavaExecutorService$$Job_1604746675.gen:13:in `run'"]}
```

If I go poking the code, I see https://github.com/logstash-plugins/logstash-output-s3/blob/master/lib/logstash/outputs/s3/uploader.rb#L36-L47
```
          begin
            obj = bucket.object(file.key)
            obj.upload_file(file.path, upload_options)
          rescue => e
            # When we get here it usually mean that S3 tried to do some retry by himself (default is 3)
            # When the retry limit is reached or another error happen we will wait and retry.
            #
            # Thread might be stuck here, but I think its better than losing anything
            # its either a transient errors or something bad really happened.
            logger.error("Uploading failed, retrying", :exception => e.class, :message => e.message, :path => file.path, :backtrace => e.backtrace)
            retry
          end
```

I think this either to not retry for non-s3 errors, and probably needs a cap on the number of retries.

I have no idea how it got to this state either.

**Comment by Sjaak01** [comment]
+1

Left my device without internet connectivity overnight for testing and the next day I found a 400MB log file full of messages that the S3 output couldn't connect. 

There should be a limit on the amount of errors or a limit on the interval these messages are generated in.

**Comment by BenoitHugonnard** [comment]
+1 Any improvements on this issue ?

**Comment by idsvandermolen** [comment]
The behaviour should probably be configurable by the user, with the default action to retry forever until it succeeds.

**Comment by kares** [comment]
this issue is specific to not retry-ing missing file (`Errno::ENOENT`) which has been resolved in **4.1.0**
(implemented at https://github.com/logstash-plugins/logstash-output-s3/pull/144)
